src/symbolize/gimli/elf.rs

```diff
@@ -43,6 +43,45 @@ impl Mapping {
         })
     }
 
+    /// On Android, shared objects can be loaded directly from a
+    /// ZIP archive. For example, an app may load a library from
+    /// `/data/app/com.example/base.apk!/lib/x86_64/mylib.so`
+    ///
+    /// `zip_offset` should be page-aligned; the dynamic linker
+    /// requires this when it loads libraries.
+    #[cfg(target_os = "android")]
+    pub fn new_android(path: &Path, zip_offset: Option<i64>) -> Option<Mapping> {
+        fn map_embedded_library(path: &Path, zip_offset: i64) -> Option<Mapping> {
+            // get path of ZIP archive (delimited by `!/`)
+            let raw_path = path.as_os_str().as_bytes();
+            let zip_path = raw_path
+                .windows(2)
+                .enumerate()
+                .find(|(_, chunk)| chunk == b"!/")
+                .map(|(index, _)| Path::new(OsStr::from_bytes(raw_path.split_at(index).0)))?;
+
+            let file = fs::File::open(zip_path).ok()?;
+            let len: usize = file.metadata().ok()?.len().try_into().ok()?;
+
+            // NOTE: we map the remainder of the entire archive instead of just the library so we don't have to determine its length
+            // NOTE: mmap will fail if `zip_offset` is not page-aligned
```

Comment:
invariants on mmap should be documented on the mmap function, no?

Reply:
Fair point, though the exact requirements are platform-dependent.

Linux requires page-alignment (at least on the [architectures][linux-x86-64] I [looked at][linux-arm64]), [Windows][win-mmap] seems to require alignment to a larger number (64K on the 4K system I checked), and I have no idea what other platforms' requirements may be.

Honestly I assumed that callers were responsible for making sure their usage of `mmap` met any platform-specific requirements, but that's not explicitly documented anywhere.

I could move the Android-specific note to the function, or just add a statement to the function that "callers must ensure that any platform-specific requirements are upheld"?

---

I should also note that, at least on Android (and I *think* Linux in general), a call with an unaligned offset is defined behaviour: the call returns `-1`, which the `mmap` wrapper translates to `None`.

Hence my decision to just call mmap and let it fail if the offset is unaligned, rather than retrieve the system's page size and ensure alignment beforehand.

[linux-x86-64]: https://elixir.bootlin.com/linux/v6.10.8/source/arch/x86/kernel/sys_x86_64.c#L83
[linux-arm64]: https://elixir.bootlin.com/linux/v6.10.8/source/arch/arm64/kernel/sys.c#L25
[win-mmap]: https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-mapviewoffile

Reply:
Just saw https://github.com/rust-lang/backtrace-rs/pull/667 - I suppose if the function will get some general documentation anyway, it might as well include a couple of notes to the effect of "on Android, (explain offset behaviour)" and "on other platforms, offset may have other requirements for safety or to avoid crashes"

Reply:
Yes, that it's safe (though very easy to cause a *failure*, but a defined one) is a feature of [POSIX mmap](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/functions/mmap.html).

It's just all the accesses after that which are racy as fuck.

Reply:
TIL it's part of POSIX (and that POSIX even publishes man pages for this kind of thing), thanks!

> It's just all the accesses after that which are racy as fuck.

Not sure I follow - you mean with other threads in the process?

Reply:
mmap is racy with every other process on the system.